File: languages/go/client/builder.go
package bloock

import (
	"context"
	"errors"

	"github.com/bloock/go-bridge/client/entity"
	"github.com/bloock/go-bridge/internal/bridge"
	"github.com/bloock/go-bridge/internal/bridge/proto"
)

type RecordBuilder struct {
	payload     interface{}
	payloadType proto.RecordTypes
	signer      *proto.Signer
	encrypter   *proto.Encrypter
}

func (b RecordBuilder) WithSigner(signer entity.Signer) RecordBuilder {
	b.signer = signer.ToProto()
	return b
}

func (b RecordBuilder) WithEncrypter(encrypter entity.Encrypter) RecordBuilder {
	b.encrypter = encrypter.ToProto()
	return b
}

func (b RecordBuilder) Build() (entity.Record, error) {
	bridgeClient := bridge.NewBloockBridge()

	var res *proto.RecordBuilderResponse
	var err error

	switch b.payloadType {
	case proto.RecordTypes_STRING:
		res, err = bridgeClient.Record().BuildRecordFromString(context.Background(), &proto.RecordBuilderFromStringRequest{
			Payload:   b.payload.(string),
			Signer:    b.signer,
			Encrypter: b.encrypter,
		})
	case proto.RecordTypes_BYTES:
		res, err = bridgeClient.Record().BuildRecordFromBytes(context.Background(), &proto.RecordBuilderFromBytesRequest{
			Payload:   b.payload.([]byte),
			Signer:    b.signer,
			Encrypter: b.encrypter,
		})
	case proto.RecordTypes_FILE:
		res, err = bridgeClient.Record().BuildRecordFromFile(context.Background(), &proto.RecordBuilderFromFileRequest{
			Payload:   b.payload.([]byte),
			Signer:    b.signer,
			Encrypter: b.encrypter,
		})
	case proto.RecordTypes_JSON:
		res, err = bridgeClient.Record().BuildRecordFromJson(context.Background(), &proto.RecordBuilderFromJSONRequest{
			Payload:   b.payload.(string),
			Signer:    b.signer,
			Encrypter: b.encrypter,
		})
	case proto.RecordTypes_HEX:
		res, err = bridgeClient.Record().BuildRecordFromHex(context.Background(), &proto.RecordBuilderFromHexRequest{
			Payload:   b.payload.(string),
			Signer:    b.signer,
			Encrypter: b.encrypter,
		})
	case proto.RecordTypes_RECORD:
		res, err = bridgeClient.Record().BuildRecordFromRecord(context.Background(), &proto.RecordBuilderFromRecordRequest{
			Payload:   b.payload.(*proto.Record),
			Signer:    b.signer,
			Encrypter: b.encrypter,
		})
	}

	if err != nil {
		return entity.Record{}, err
	}

	if res.Error != nil {
		return entity.Record{}, errors.New(res.Error.Message)
	}

	return entity.NewRecordFromProto(res.Record), nil
}

func NewRecordBuilderFromRecord(record entity.Record) RecordBuilder {
	return RecordBuilder{
		payload:     record.ToProto(),
		payloadType: proto.RecordTypes_RECORD,
	}
}

func NewRecordBuilderFromString(str string) RecordBuilder {
	return RecordBuilder{
		payload:     str,
		payloadType: proto.RecordTypes_STRING,
	}
}

func NewRecordBuilderFromHex(hex string) RecordBuilder {
	return RecordBuilder{
		payload:     hex,
		payloadType: proto.RecordTypes_HEX,
	}
}

func NewRecordBuilderFromJSON(json string) RecordBuilder {
	return RecordBuilder{
		payload:     json,
		payloadType: proto.RecordTypes_JSON,
	}
}

func NewRecordBuilderFromFile(file_bytes []byte) RecordBuilder {
	return RecordBuilder{
		payload:     file_bytes,
		payloadType: proto.RecordTypes_FILE,
	}
}

func NewRecordBuilderFromBytes(bytes []byte) RecordBuilder {
	return RecordBuilder{
		payload:     bytes,
		payloadType: proto.RecordTypes_BYTES,
	}
}

File: languages/java/lib/src/main/java/bloock/sdk/java/builder/Builder.java
package bloock.sdk.java.builder;

import com.google.protobuf.ByteString;

import bloock.sdk.java.bridge.Bridge;
import bloock.sdk.java.bridge.proto.RecordOuterClass;
import bloock.sdk.java.bridge.proto.RecordOuterClass.RecordBuilderFromBytesRequest;
import bloock.sdk.java.bridge.proto.RecordOuterClass.RecordBuilderFromFileRequest;
import bloock.sdk.java.bridge.proto.RecordOuterClass.RecordBuilderFromHexRequest;
import bloock.sdk.java.bridge.proto.RecordOuterClass.RecordBuilderFromJSONRequest;
import bloock.sdk.java.bridge.proto.RecordOuterClass.RecordBuilderFromRecordRequest;
import bloock.sdk.java.bridge.proto.RecordOuterClass.RecordBuilderFromStringRequest;
import bloock.sdk.java.bridge.proto.RecordOuterClass.RecordBuilderResponse;
import bloock.sdk.java.bridge.proto.RecordOuterClass.RecordTypes;
import bloock.sdk.java.bridge.proto.Shared.Error;
import bloock.sdk.java.entity.Decrypter;
import bloock.sdk.java.entity.Encrypter;
import bloock.sdk.java.entity.Record;
import bloock.sdk.java.entity.Signer;

public class Builder {
    Object payload;
    RecordTypes type;
    RecordOuterClass.Signer signer;
    RecordOuterClass.Encrypter encrypter;
    RecordOuterClass.Decrypter decrypter;

    Builder(Object payload, RecordTypes type) {
        this.payload = payload;
        this.type = type;
    }

    public static Builder fromRecord(Record record) {
        return new Builder(record.toProto(), RecordTypes.RECORD);
    }

    public static Builder fromString(String str) {
        return new Builder(str, RecordTypes.STRING);
    }

    public static Builder fromHex(String str) {
        return new Builder(str, RecordTypes.HEX);
    }

    public static Builder fromJson(String json) {
        return new Builder(json, RecordTypes.JSON);
    }

    public static Builder fromFile(byte[] file) {
        return new Builder(file, RecordTypes.FILE);
    }

    public static Builder fromBytes(byte[] bytes) {
        return new Builder(bytes, RecordTypes.BYTES);
    }

    public Builder withSigner(Signer signer) {
        this.signer = signer.toProto();
        return this;
    }

    public Builder withEncrypter(Encrypter encrypter) {
        this.encrypter = encrypter.toProto();
        return this;
    }

    public Builder withDecrypter(Decrypter decrypter) {
        this.decrypter = decrypter.toProto();
        return this;
    }

    public Record build() throws Exception {
        Bridge bridge = new Bridge();

        RecordBuilderResponse response;

        switch (this.type) {
            case BYTES: {
                RecordBuilderFromBytesRequest.Builder builder = RecordBuilderFromBytesRequest
                        .newBuilder()
                        .setPayload(ByteString.copyFrom((byte[]) this.payload));

                if (this.signer != null) {
                    builder.setSigner(this.signer);
                }
                if (this.encrypter != null) {
                    builder.setEncrypter(this.encrypter);
                }
                if (this.decrypter != null) {
                    builder.setDecrypter(this.decrypter);
                }

                response = bridge.getRecord().buildRecordFromBytes(builder.build());
                break;
            }
            case FILE: {
                RecordBuilderFromFileRequest.Builder builder = RecordBuilderFromFileRequest
                        .newBuilder()
                        .setPayload(ByteString.copyFrom((byte[]) this.payload));

                if (this.signer != null) {
                    builder.setSigner(this.signer);
                }
                if (this.encrypter != null) {
                    builder.setEncrypter(this.encrypter);
                }
                if (this.decrypter != null) {
                    builder.setDecrypter(this.decrypter);
                }

                response = bridge.getRecord().buildRecordFromFile(builder.build());
                break;
            }
            case HEX: {
                RecordBuilderFromHexRequest.Builder builder = RecordBuilderFromHexRequest
                        .newBuilder()
                        .setPayload((String) this.payload);

                if (this.signer != null) {
                    builder.setSigner(this.signer);
                }
                if (this.encrypter != null) {
                    builder.setEncrypter(this.encrypter);
                }
                if (this.decrypter != null) {
                    builder.setDecrypter(this.decrypter);
                }

                response = bridge.getRecord().buildRecordFromHex(builder.build());
                break;
            }
            case JSON: {
                RecordBuilderFromJSONRequest.Builder builder = RecordBuilderFromJSONRequest
                        .newBuilder()
                        .setPayload((String) this.payload);

                if (this.signer != null) {
                    builder.setSigner(this.signer);
                }
                if (this.encrypter != null) {
                    builder.setEncrypter(this.encrypter);
                }
                if (this.decrypter != null) {
                    builder.setDecrypter(this.decrypter);
                }

                response = bridge.getRecord().buildRecordFromJson(builder.build());
                break;
            }
            case RECORD: {
                RecordBuilderFromRecordRequest.Builder builder = RecordBuilderFromRecordRequest
                        .newBuilder()
                        .setPayload((RecordOuterClass.Record) this.payload);

                if (this.signer != null) {
                    builder.setSigner(this.signer);
                }
                if (this.encrypter != null) {
                    builder.setEncrypter(this.encrypter);
                }
                if (this.decrypter != null) {
                    builder.setDecrypter(this.decrypter);
                }

                response = bridge.getRecord().buildRecordFromRecord(builder.build());
                break;
            }
            case STRING: {
                RecordBuilderFromStringRequest.Builder builder = RecordBuilderFromStringRequest
                        .newBuilder()
                        .setPayload((String) this.payload);

                if (this.signer != null) {
                    builder.setSigner(this.signer);
                }
                if (this.encrypter != null) {
                    builder.setEncrypter(this.encrypter);
                }
                if (this.decrypter != null) {
                    builder.setDecrypter(this.decrypter);
                }

                response = bridge.getRecord().buildRecordFromString(builder.build());
                break;
            }
            default:
                throw new Exception("Invalid type");
        }

        if (response.getError() != Error.getDefaultInstance()) {
            throw new Exception(response.getError().getMessage());
        }

        return Record.fromProto(response.getRecord());
    }
}
